chaincode/contracts/AssetManager.sol

```diff
@@ -0,0 +1,190 @@
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "./Enums.sol";
+
+contract AssetManager is Ownable {
+    event DepositSuccesful(address user, uint256 amount, uint256 balance);
+    // event DepositFailed(
+    //     address depositor,
+    //     uint256 amount,
+    //     uint256 balance,
+    //     string reason
+    // );
+    event WithdrawalSuccesful(address user, uint256 amount, uint256 balance);
+    event WithdrawalFailed(
+        address user,
+        uint256 amount,
+        uint256 balance,
+        string reason
+    );
+    event AuthorizationSuccesful(address user, uint256 newLimit, uint256 userBalance);
+    event AuthorizationFailed(
+        address depositor,
+        uint256 userBalance,
+        uint256 limit,
+        string reason
+    );
+    event SendSuccesful(address from, address to, uint256 amount, uint256 newBalance, uint256 newLimit);
+    event SendFailed(
+        address from,
+        address to,
+        uint256 amount,
+        uint256 balance,
+        uint256 limit,
+        string reason
+    );
+
+
+    mapping(address => uint256) public userBalances;
+    mapping(address => uint256) public userAuthorizations;
+
+    constructor() {}
+
+    function getUserBalance (address user) public view returns (uint256 balance){
+        return this.userBalances(address(user));
+    }
+
+    function getUserAuthorization (address user) public view returns (uint256 authorization){
+        return this.userAuthorizations(address(user));
+    }
+
+    // authorize
+    // send
+    // transfer
+    // pause
+
+    function deposit() public payable {
+        userBalances[address(msg.sender)] += msg.value;
+        // update the userBalance
+        emit DepositSuccesful(
+            msg.sender,
+            msg.value,
+            userBalances[address(msg.sender)]
+        );
+    }
+
+        function withdraw(uint256 amount) public payable {
+        uint256 balance =  userBalances[address(msg.sender)];
+        // if zero is passed withdraw all funds
+        if (amount == 0){ amount = balance; }
+        // check msg.senders balance
+        if (amount > balance) {
+            emit WithdrawalFailed(
+                msg.sender,
+                amount,
+                balance,
+                "Insufficient Locked Funds to Withdraw "
+            );
+            return;
+        }
+
+        require(amount <= balance);
+        // withdraw funds from the contract (update userBalance before transfer to protect from reentracy attack)
+        uint256 newBalance = balance - amount;
+        userBalances[address(msg.sender)] = newBalance;
+        payable(msg.sender).transfer(amount);
+
+        // update the authorized amount.
+        uint256 currentLimit = getUserAuthorization(msg.sender);
+        if (currentLimit > newBalance ) { authorize(newBalance); }
```

Comment:
Should authorize the difference, not newBalance

Reply:
See Clarifications above

Reply:
Withdrawal (and deposits) no longer effect approvals
i.e. you can have multiple approvals which can be greater than the amount deposited.